esmvaltool/diag_scripts/lifetime/lifetime.py

```diff
@@ -0,0 +1,1498 @@
+    alpha: 0.5`` are used. Use ``gridline_kwargs: false`` to not show grid
+    lines.
+legend_kwargs: dict, optional
+    Optional keyword arguments for :func:`matplotlib.pyplot.legend`. Use
+    ``legend_kwargs: false`` to not show legends.
+log_x: bool, optional (default: False)
+    Use logarithmic X-axis. Note that for the logarithmic x axis tickmarks are
+    set so that minor tickmarks show up. Setting of individual tickmarks by
+    pyplot_kwargs is not recommended in this case.
+log_y: bool, optional (default: True)
+    Use logarithmic Y-axis.
+plot_kwargs: dict, optional
+    Optional keyword arguments for :func:`iris.plot.plot`. Dictionary keys are
+    elements identified by ``facet_used_for_labels`` or ``default``, e.g.,
+    ``CMIP6`` if ``facet_used_for_labels: project`` or ``historical`` if
+    ``facet_used_for_labels: exp``. Dictionary values are dictionaries used as
+    keyword arguments for :func:`iris.plot.plot`. String arguments can include
+    facets in curly brackets which will be derived from the corresponding
+    dataset, e.g., ``{project}``, ``{short_name}``, ``{exp}``. Examples:
+    ``default: {linestyle: '-', label: '{project}'}, CMIP6: {color: red,
+    linestyle: '--'}, OBS: {color: black}``.
+pyplot_kwargs: dict, optional
+    Optional calls to functions of :mod:`matplotlib.pyplot`. Dictionary keys
+    are functions of :mod:`matplotlib.pyplot`. Dictionary values are used as
+    single argument for these functions. String arguments can include facets in
+    curly brackets which will be derived from the datasets plotted in the
+    corresponding plot, e.g., ``{short_name}``, ``{exp}``. Facets like
+    ``{project}`` that vary between the different datasets will be transformed
+    to something like  ``ambiguous_project``. Examples: ``title: 'Awesome Plot
+    of {long_name}'``, ``xlabel: '{short_name}'``, ``xlim: [0, 5]``.
+show_y_minor_ticklabels: bool, optional (default: False)
+    Show tick labels for the minor ticks on the Y axis.
+
+.. hint::
+
+   Extra arguments given to the recipe are ignored, so it is safe to use yaml
+   anchors to share the configuration of common arguments with other monitor
+   diagnostic script.
+
+"""
+import logging
+from copy import deepcopy
+from pathlib import Path
+from pprint import pformat
+
+import sys
+# import cartopy.crs as ccrs
+import iris
+import matplotlib as mpl
+import matplotlib.pyplot as plt
+import numpy as np
+import seaborn as sns
+from iris.analysis.calculus import cube_delta
+# from iris.analysis.maths import log, exp
+from iris.coord_categorisation import add_year
+from iris.util import broadcast_to_shape
+import iris.common
+# from iris.coords import AuxCoord
+from matplotlib.gridspec import GridSpec
+from matplotlib.ticker import FormatStrFormatter, LogLocator, NullFormatter
+# from sklearn.metrics import r2_score
+from scipy.constants import N_A, R, G
+
+from esmvalcore.cmor._fixes.shared import add_model_level
```

Comment:
I can't find this function in the current ESMValCore version. Also, please do not import private objects (i.e., those that start with with a `_`) in your code, as those might change without any prior notice.

Reply:
Yes, I added this function to my developer version of ESMValCore (branch `cmor-4-lifetime`). Maybe it is not an appropriate place for this function. I need the number of the model level and would like to add it by the preprocessor. However, it gets lost during the diagnostic. I am unsure if I should leave it in the ESMValCore (at some other point) or put it into the lifetime diagnostic itself. What do you think? I thought it might be useful for other applications as well.

Reply:
Ah, I see, that makes sense. If you use the function in ESMValCore, it's fine to leave it there. You can then just add a public version of it [here](https://github.com/ESMValGroup/ESMValCore/blob/cmor-4-lifetime/esmvalcore/cmor/fixes.py), which you can then import via


```py
from esmvalcore.cmor.fixes import add_model_level
```

If you don't use it in ESMValCore, it's better to just put it in the diagnostic.